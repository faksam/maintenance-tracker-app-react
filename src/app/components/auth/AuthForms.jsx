import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import isEmpty from 'is-empty';

import SignInTabForm from './SignInTabForm';
import SignUpTabForm from './SignUpTabForm';

import signInAction from '../../actions/signin.actions';
import signUpAction from '../../actions/signup.actions';
import UserInputValidation from '../../validator/authValidation';

/**
 * @class AuthForm
 *
 * @description Comment Form Component
 * @extends {Component}
 *
 * @param {*} event
 *
 * @param {*} formName
 * @param {*} userInput
 */
export class AuthForm extends Component {
  /**
   * @description - Class Constructor
   *
   * @param {*} props
   */
  constructor(props) {
    super(props);

    this.state = {
      errors: {},
      email: '',
      password: '',
      fullname: '',
      confirmPassword: '',
      accountEmail: '',
      accountPassword: '',
      signInFormTabContent: 'tabcontent',
      signUpFormTabContent: 'tabcontent-display',
      signInTab: 'tablinks sign-tab',
      signUpTab: 'tablinks sign-tab active',
      errorDivClass: 'display-none',
    };
  }

  static getDerivedStateFromProps(nextProps) {
    const { auth, history, user } = nextProps;
    if (auth && user.role === 'User') {
      history.push('/dashboard');
      return null;
    } else if (auth && user.role === 'Admin') {
      history.push('/admin');
    }
    return null;
  }

  handleChange = (event) => {
    const { errors } = this.state;
    if (errors[event.target.name]) {
      const newErrors = Object.assign({}, errors);
      delete newErrors[event.target.id];
      this.setState({
        [event.target.id]: event.target.value, errors: newErrors
      });
    } else {
      this.setState({
        [event.target.id]: event.target.value
      });
    }
  }

  handleSignIn = (event) => {
    event.preventDefault();
    this.setState({
      errors: {}
    });
    const { accountEmail, accountPassword } = this.state;
    const { signInUser } = this.props;
    const userInput = {
      email: accountEmail,
      password: accountPassword
    };
    const { errors, isValid } = UserInputValidation.signInInputValidation(userInput);

    if (isValid) {
      signInUser(userInput);
    } else {
      this.setState({
        errors,
        errorDivClass: 'display-block',
      });
    }
  }

  handleSignUp = (event) => {
    event.preventDefault();
    this.setState({
      errors: {}
    });
    const {
      fullname, email, password, confirmPassword
    } = this.state;
    const { signUpUser } = this.props;
    const userInput = {
      fullname,
      email,
      password,
      password_confirmation: confirmPassword
    };
    const { errors, isValid } = UserInputValidation.signUpInputValidation(userInput);

    const validatedInput = {
      fullName: fullname,
      email,
      password,
      confirmPassword
    };
    if (isValid) {
      signUpUser(validatedInput);
    } else {
      this.setState({
        errors,
        errorDivClass: 'display-block',
      });
    }
  }

  handleCloseError = () => {
    this.setState({
      errors: {}
    });
  };

  displayForm = (event) => {
    event.preventDefault();
    if (event.target.id === 'signUpTab') {
      this.setState({
        signUpTab: 'tablinks sign-tab active',
        signUpFormTabContent: 'tabcontent-display',
        signInFormTabContent: 'tabcontent',
        signInTab: 'tablinks sign-tab',
        errors: {}
      });
    } else {
      this.setState({
        signUpTab: 'tablinks sign-tab',
        signInTab: 'tablinks sign-tab active',
        signInFormTabContent: 'tabcontent-display',
        signUpFormTabContent: 'tabcontent',
        errors: {}
      });
    }
  };

  /**
   * @memberof AuthForm
   *
   * @returns {JSX} JSX representation of component
   *
   * @description Render the JSX template
   */
  render() {
    const {
      errors,
      email,
      password,
      fullname,
      confirmPassword,
      accountPassword,
      accountEmail,
      signInFormTabContent,
      signUpFormTabContent,
      signUpTab,
      signInTab,
      errorDivClass,
    } = this.state;

    const {
      error,
    } = this.props;
    return (
      <div className="row">
        <div className="tab col col-1">
          <button
            id="signUpTab"
            type="button"
            onClick={this.displayForm}
            className={signUpTab}
          >
            Sign Up
          </button>
          <button
            id="signInTab"
            type="button"
            onClick={this.displayForm}
            className={signInTab}
          >
            Sign In
          </button>
        </div>
        <div id="loginErrorDiv">
          <p id="loginErrorMessage" className="hide-div" />
        </div>
        {
          !isEmpty(error) && (<p className="danger">{error.message}</p>)
        }
        <div id="signInTabForm" className={signInFormTabContent}>
          <SignInTabForm
            errors={errors}
            errorDivClass={errorDivClass}
            handleChange={this.handleChange}
            handleSignIn={this.handleSignIn}
            accountEmail={accountEmail}
            accountPassword={accountPassword}
          />
        </div>
        <div id="signUpTabForm" className={signUpFormTabContent}>
          <SignUpTabForm
            errors={errors}
            errorDivClass={errorDivClass}
            handleChange={this.handleChange}
            handleSignUp={this.handleSignUp}
            fullname={fullname}
            email={email}
            password={password}
            confirmPassword={confirmPassword}
          />
        </div>
      </div>
    );
  }
}

AuthForm.propTypes = {
  error: PropTypes.instanceOf(Object),
  signInUser: PropTypes.func,
  signUpUser: PropTypes.func,
};

const mapStateToProps = state => ({
  auth: state.auth.isAuthenticated,
  user: state.auth.user,
  error: state.auth.error
});

const matchDispatchToProps = dispatch => bindActionCreators({
  signInUser: signInAction,
  signUpUser: signUpAction
}, dispatch);

export default connect(mapStateToProps, matchDispatchToProps)(AuthForm);
