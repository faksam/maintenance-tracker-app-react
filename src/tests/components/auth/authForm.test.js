import React from 'react';
import { shallow } from 'enzyme';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import AuthForms from '../../../app/components/auth/AuthForms';

const props = {
  fullname: 'fakunlesamuel@gmail.com',
  email: 'fakunlesamuel@gmail.com',
  password: 'Lifeisarace',
  confirmPassword: 'Lifeisarace',
  handleSignUp: jest.fn(),
  handleChange: jest.fn(),
};
const initialState = {
  auth: {
    isAuthenticated: true,
    error: {},
    user: {
      username: 'joeeasy',
      email: 'joeeasy@gmail.com',
      role: 'User',
      id: '6bf54873-6033-4ab0-855d-336a34e1dfbc',
      iat: 1540985284931
    }
  },

};
const mockStore = configureMockStore([thunk]);
const store = mockStore(initialState);

describe('SignInTabForm Component', () => {
  it('should render without throwing an error', () => {
    const wrapper = shallow(<AuthForms store={store} {...props} />);
    expect(wrapper.exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });
  it('should set state to props', () => {
    const wrapper = shallow(<AuthForms store={store} {...props} />);
    expect(wrapper.state()).toEqual({});
  });
  // it('should not create an empty article', () => {
  //   const wrapper = shallow(<CommentForm {...props} />);
  //   const instance = wrapper.instance();
  //   const event = {
  //     preventDefault: jest.fn()
  //   };
  //   wrapper.setState({ body: '' });
  //   expect(instance.isValid()).toBe(false);
  //   instance.onSubmit(event);
  // });
  // it('should a new comment', () => {
  //   const wrapper = shallow(<CommentForm {...props} />);
  //   const instance = wrapper.instance();
  //   const event = {
  //     preventDefault: jest.fn()
  //   };
  //   wrapper.setState({ body: 'Samuel is a boy' });
  //   expect(instance.isValid()).toBe(true);
  //   instance.onSubmit(event);
  // });
  // it('should close error alert', () => {
  //   const wrapper = shallow(<CommentForm {...props} />);
  //   const instance = wrapper.instance();
  //   instance.handleCloseError();
  // });
  // it('should set comment body when comment form body changes', () => {
  //   const wrapper = shallow(<CommentForm {...props} />);
  //   const event = {
  //     preventDefault: jest.fn(),
  //     target: {
  //       id: 'body',
  //       value: 'This is a comment'
  //     }
  //   };
  //   wrapper.setState({ errors: { [event.target.id]: 'mock' } });
  //   wrapper.instance().onChange(event);
  //   expect(wrapper.instance().state[event.target.id])
  //     .toBe(event.target.value);
  // });
  // it('should set state to props', () => {
  //   const wrapper = shallow(<ConnectedCommentForm store={store} {...props} />);
  //   expect(wrapper.state()).toEqual({});
  // });
});
