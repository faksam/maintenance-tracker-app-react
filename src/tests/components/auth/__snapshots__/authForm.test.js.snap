// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInTabForm Component should render without throwing an error 1`] = `
<AuthForm
  auth={true}
  confirmPassword="Lifeisarace"
  email="fakunlesamuel@gmail.com"
  error={Object {}}
  fullname="fakunlesamuel@gmail.com"
  handleChange={[MockFunction]}
  handleSignUp={[MockFunction]}
  password="Lifeisarace"
  signInUser={[Function]}
  signUpUser={[Function]}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  storeSubscription={
    Subscription {
      "listeners": Object {
        "clear": [Function],
        "get": [Function],
        "notify": [Function],
        "subscribe": [Function],
      },
      "onStateChange": [Function],
      "parentSub": undefined,
      "store": Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      },
      "unsubscribe": [Function],
    }
  }
  user={
    Object {
      "email": "joeeasy@gmail.com",
      "iat": 1540985284931,
      "id": "6bf54873-6033-4ab0-855d-336a34e1dfbc",
      "role": "User",
      "username": "joeeasy",
    }
  }
/>
`;
